### PR TITLE
En_Water_Effect

### PR DESCRIPTION
**Pull request by Kelebek1**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

### CODE REVIEW

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c [hunk 1]
```diff
@@ -28,30 +46,672 @@ const ActorInit En_Water_Effect_InitVars = {
...
+            if (Rand_ZeroOne() < 0.5f) {
+                Actor_UpdateBgCheckInfo(globalCtx, &this->actor, 20.0f, 10.0f, 40.0f, 4);
+                sp88.x = randPlusMinusPoint5Scaled(50.0f) + this->actor.world.pos.x;
+                sp88.y = this->actor.world.pos.y;
+                sp88.z = randPlusMinusPoint5Scaled(50.0f) + this->actor.world.pos.z;
+                if ((this->actor.world.pos.y + this->actor.depthInWater) <= this->actor.floorHeight) {
+                    phi_v1 = 0;
+                } else {
+                    phi_v1 = 1;
+                    this->actor.floorHeight = (this->actor.world.pos.y + this->actor.depthInWater);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:170]:
```suggestion
                    this->actor.floorHeight = this->actor.world.pos.y + this->actor.depthInWater;
```

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c [hunk 2]
```diff
@@ -28,30 +46,672 @@ const ActorInit En_Water_Effect_InitVars = {
...
+        this->unk_DC6 = Rand_ZeroFloat(150.0f) + 100.0f;
+    } else if (!func_801690CC(globalCtx)) {
+        this->unk_DC4++;
+        if (this->unk_DC6 == 0) {
+            this->unk_DC6 = Rand_ZeroFloat(150.0f) + 100.0f;
+            Actor_UpdateBgCheckInfo(globalCtx, &this->actor, 20.0f, 10.0f, 40.0f, 4);
+            sp74.x = randPlusMinusPoint5Scaled(50.0f) + this->actor.world.pos.x;
+            sp74.y = this->actor.world.pos.y;
+            sp74.z = randPlusMinusPoint5Scaled(50.0f) + this->actor.world.pos.z;
+            func_80A599E8(this, &sp74, 0U);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:428]:
```suggestion
            func_80A599E8(this, &sp74, 0);
```
Needed? Could you also mess around with the prototype of `func_80A599E8`? Applies to other cases of `func_80A599E8` below

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c [hunk 3]
```diff
@@ -28,30 +46,672 @@ const ActorInit En_Water_Effect_InitVars = {
...
+
+            if (ptr->unk_2A >= 2) {
+                Matrix_NormalizeXYZ(&globalCtx->mf_187FC);
+            } else {
+                Matrix_RotateY(func_800DFC68(GET_ACTIVE_CAM(globalCtx)), MTXMODE_APPLY);
+            }
+
+            Matrix_Scale(ptr->unk_2C.x, ptr->unk_2C.y, 1.0f, MTXMODE_APPLY);
+
+            if ((i & 1) != 0) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:555]:
```suggestion
            if ((i % 2) != 0) {
```
If it matches

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c [hunk 4]
```diff
@@ -28,30 +46,672 @@ const ActorInit En_Water_Effect_InitVars = {
...
+                Matrix_InsertYRotation_f(M_PI, MTXMODE_APPLY);
+            }
+
+            gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx),
+                      G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+            gSPDisplayList(POLY_XLU_DISP++, D_060043E8);
+
+            if ((ptr->unk_2A & 1) == 0) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:563]:
```suggestion
            if ((ptr->unk_2A % 2) == 0) {
```
If it matches

**Review by engineer124 — COMMENTED** [review]
Nice, some small comments

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c (hunk 3 above)
Comment (Kelebek1) [reply src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:555]: Doesn't

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.h [hunk 5]
```diff
@@ -5,9 +5,42 @@
...
+} EnWaterEffectStruct; // size = 0x40
+
+enum {
+    /*     1 */ ENWATEREFFECT_1 = 1,
+    /* 0x309 */ ENWATEREFFECT_777 = 777,
+    /* 0x30A */ ENWATEREFFECT_778,
+    /* 0x30B */ ENWATEREFFECT_779,
+    /* 0x30C */ ENWATEREFFECT_780,
+
+};
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.h:28]:
```suggestion
    /* 0x30C */ ENWATEREFFECT_780,
};
```

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.h [hunk 6]
```diff
@@ -5,9 +5,42 @@
...
     /* 0x000 */ Actor actor;
-    /* 0x144 */ char unk_144[0xCF8];
+    /* 0x144 */ EnWaterEffectStruct unk_144[50];
+    /* 0xDC4 */ s16 unk_DC4;
+    /* 0xDC6 */ s16 unk_DC6;
+    /* 0xDC8 */ Vec3f unk_DC8[5];
+    /* 0xE04 */ UNK_TYPE1 unkE04[4];
+    /* 0xE04 */ f32 unk_E08[4];
+    /* 0xE18 */ f32 unk_E18[4];
+    /* 0xE28 */ UNK_TYPE1 unkE28[4];
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.h:39]:
```suggestion
    /* 0xDC8 */ Vec3f unk_DC8[5];
    /* 0xE04 */ UNK_TYPE1 unk_E04[4];
    /* 0xE04 */ f32 unk_E08[4];
    /* 0xE18 */ f32 unk_E18[4];
    /* 0xE28 */ UNK_TYPE1 unk_E28[4];
```

**Review by AngheloAlf — COMMENTED** [review]
Some minor style requests

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c [hunk 7]
```diff
@@ -28,30 +33,672 @@ const ActorInit En_Water_Effect_InitVars = {
...
+    OPEN_DISPS(gfxCtx);
+
+    func_8012C2DC(globalCtx->state.gfxCtx);
+
+    for (i = 0; i < ARRAY_COUNT(this->unk_144) / 2; i++, ptr++) {
+        if ((ptr->unk_00 == 1) || (ptr->unk_00 == 2)) {
+            if (!phi_s4) {
+                POLY_XLU_DISP = Gfx_CallSetupDL(POLY_XLU_DISP, 0);
+
+                gSPSegment(POLY_XLU_DISP++, 0x08, Lib_SegmentedToVirtual(D_0408DBE0));
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:287]: `D_0408DBE0` is from gameplay keep. Should be able to include and use the symbol.

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c [hunk 8]
```diff
@@ -28,30 +33,672 @@ const ActorInit En_Water_Effect_InitVars = {
...
+
+    phi_s4 = false;
+    ptr = backupPtr;
+
+    for (i = 0; i < ARRAY_COUNT(this->unk_144) / 2; i++, ptr++) {
+        if (ptr->unk_00 == 3) {
+            if (!phi_s4) {
+                func_8012C448(gfxCtx);
+
+                gSPSegment(POLY_XLU_DISP++, 0x08, Lib_SegmentedToVirtual(D_0408DBE0));
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:319]: Same here

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c [hunk 9]
```diff
@@ -28,30 +33,672 @@ const ActorInit En_Water_Effect_InitVars = {
...
+            gSPDisplayList(POLY_XLU_DISP++, object_water_effect_DL_0043E8);
+
+            if ((ptr->unk_2A & 1) == 0) {
+                Matrix_InsertTranslation(ptr->unk_04.x, ptr->unk_04.y + 5.0f, ptr->unk_04.z, MTXMODE_NEW);
+                Matrix_InsertXRotation_s(ptr->unk_28, MTXMODE_APPLY);
+                Matrix_Scale(ptr->unk_2C.z, ptr->unk_2C.z, ptr->unk_2C.z, MTXMODE_APPLY);
+
+                gSPMatrix(POLY_OPA_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx),
+                          G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+                gSPDisplayList(POLY_OPA_DISP++, D_0406AB30);
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:557]: `D_0406AB30` is from gameplay keep. Should be able to include and use the symbol.

File: src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c [hunk 10]
```diff
@@ -28,30 +33,672 @@ const ActorInit En_Water_Effect_InitVars = {
...
+        func_8012C2DC(globalCtx->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Water_Effect/EnWaterEffect_Draw.s")
+        for (i = 0; i < ARRAY_COUNT(this->unk_144) / 2; i++, ptr++) {
+            if (ptr->unk_00 == 3) {
+                if (!phi_s4) {
+                    func_8012C448(globalCtx->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Water_Effect/func_80A599E8.s")
+                    gSPSegment(POLY_XLU_DISP++, 0x08, Lib_SegmentedToVirtual(D_0408DBE0));
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Water_Effect/z_en_water_effect.c:683]: Same here

**Review by hensldm — APPROVED** [review]

**Review by AngheloAlf — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**